docs/source/library-user-guide/api-health.md

```diff
@@ -69,3 +69,32 @@ For example:
 Deprecated methods will remain in the codebase for a period of 6 major versions or 6 months, whichever is longer, to provide users ample time to transition away from them.
 
 Please refer to [DataFusion releases](https://crates.io/crates/datafusion/versions) to plan ahead API migration
+
+## Migration Guidelines
+
+To ensure smooth upgrades and maintain application stability, the following guidelines must be followed for changes involving:
+
+- Public API changes
+- Introducing deprecated methods
+- Removal of obsolete methods
```

Comment:
I personally don't think the last two items are worth documenting -- deprecated methods should have already been removed so there is no need for a contributor to try and document anything when removing them

I would also like us to get more in the habit of deprecating rather than removing methods which will make the last item irrelevant.

Reply:
Deprecated functions are self explanatory if deprecation guidelines followed :) 

But one pros for documenting it in migration plan is the user is able to identify amount of migration work needed just by looking into the guide for a specific version. 

For example https://spark.apache.org/docs/3.5.3/core-migration-guide.html#upgrading-from-core-34-to-35

Reply:
> I personally don't think the last two items are worth documenting

agreed

> But one pros for documenting it in migration plan is the user is able to identify amount of migration work needed just by looking into the guide for a specific version.

otoh it adds burden for development, and sets wrong incentives, discouraging from deprecating methods and removing old code

If this is indeed useful, let's maybe automate pulling of new deprecations, rather than impose a new tax

Reply:
Automatic deprecation doc generation sounds good. We can think of it later.
I'm not sure about removing it completely as it tracks API evolution for the user. 

There would be always a balance between development work and migration support. From recent days we understood user migration is important for them